Cannot install jsonnet gem

I cannot install the gem with Ruby3.3.1.

```console
❯ dpkg -l | grep jsonnet
ii  jsonnet                                  0.20.0+ds-1+b2                                          amd64        data templating language
ii  libjsonnet-dev                           0.20.0+ds-1+b2                                          amd64        data templating language (devel)
ii  libjsonnet0                              0.20.0+ds-1+b2                                          amd64        data templating language (lib)
❯ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux trixie/sid
Release:        n/a
Codename:       trixie
❯ ruby -v
ruby 3.3.1 (2024-04-23 revision c56cd86388) [x86_64-linux]
❯ gcc --version
gcc (Debian 14.2.0-1) 14.2.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```console
❯ gem i jsonnet
Building native extensions. This could take a while...
ERROR:  Error installing jsonnet:
        ERROR: Failed to build gem native extension.

    current directory: /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3/ext/jsonnet
/home/kenji/.local/share/mise/installs/ruby/3.3.1/bin/ruby extconf.rb
Building jsonnet using packaged libraries.
Using mini_portile version 2.8.7
Downloading v0.18.0.tar.gz (100%)
Extracting v0.18.0.tar.gz into tmp/x86_64-linux-gnu/ports/jsonnet/v0.18.0... OK
Running 'compile' for jsonnet v0.18.0... ERROR. Please review logs to see what happened:
----- contents of '/home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3/ext/jsonnet/tmp/x86_64-linux-gnu/ports/jsonnet/v0.18.0/compile.log' -----
((od -v -Anone -t u1 stdlib/std.jsonnet \
        | tr " " "\n" \
        | grep -v "^$" \
        | tr "\n" "," ) && echo "0") > core/std.jsonnet.h
echo >> core/std.jsonnet.h
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
g++  core/desugarer.o core/formatter.o core/lexer.o core/libjsonnet.o core/parser.o core/pass.o core/static_analysis.o core/string_utils.o core/vm.o third_party/md5/md5.o third_party/rapidyaml/rapidyaml/src/c4/y
ln -sf libjsonnet.so.0.18.0 libjsonnet.so.0
ln -sf libjsonnet.so.0 libjsonnet.so
g++ -c -g -O3 -Wall -Wextra -Woverloaded-virtual -pedantic -std=c++0x -fPIC -Iinclude -Ithird_party/md5 -Ithird_party/json -Ithird_party/rapidyaml/rapidyaml/src/ -Ithird_party/rapidyaml/rapidyaml/ext/c4core/src/
In file included from cpp/libjsonnet++.cpp:17:
include/libjsonnet++.h:49:22: error: ‘uint32_t’ has not been declared
   49 |     void setMaxStack(uint32_t depth);
      |                      ^~~~~~~~
include/libjsonnet++.h:53:26: error: ‘uint32_t’ has not been declared
   53 |     void setGcMinObjects(uint32_t objects);
      |                          ^~~~~~~~
include/libjsonnet++.h:63:22: error: ‘uint32_t’ has not been declared
   63 |     void setMaxTrace(uint32_t lines);
      |                      ^~~~~~~~
cpp/libjsonnet++.cpp:41:6: error: variable or field ‘setMaxStack’ declared void
   41 | void Jsonnet::setMaxStack(uint32_t depth)
      |      ^~~~~~~
cpp/libjsonnet++.cpp:41:27: error: ‘uint32_t’ was not declared in this scope
   41 | void Jsonnet::setMaxStack(uint32_t depth)
      |                           ^~~~~~~~
cpp/libjsonnet++.cpp:18:1: note: ‘uint32_t’ is defined in header ‘<cstdint>’; this is probably fixable by adding ‘#include <cstdint>’
   17 | #include "libjsonnet++.h"
  +++ |+#include <cstdint>
   18 |
cpp/libjsonnet++.cpp:46:6: error: variable or field ‘setGcMinObjects’ declared void
   46 | void Jsonnet::setGcMinObjects(uint32_t objects)
      |      ^~~~~~~
cpp/libjsonnet++.cpp:46:31: error: ‘uint32_t’ was not declared in this scope
   46 | void Jsonnet::setGcMinObjects(uint32_t objects)
      |                               ^~~~~~~~
cpp/libjsonnet++.cpp:46:31: note: ‘uint32_t’ is defined in header ‘<cstdint>’; this is probably fixable by adding ‘#include <cstdint>’
cpp/libjsonnet++.cpp:66:6: error: variable or field ‘setMaxTrace’ declared void
   66 | void Jsonnet::setMaxTrace(uint32_t lines)
      |      ^~~~~~~
cpp/libjsonnet++.cpp:66:27: error: ‘uint32_t’ was not declared in this scope
   66 | void Jsonnet::setMaxTrace(uint32_t lines)
      |                           ^~~~~~~~
cpp/libjsonnet++.cpp:66:27: note: ‘uint32_t’ is defined in header ‘<cstdint>’; this is probably fixable by adding ‘#include <cstdint>’
make: *** [Makefile:167: cpp/libjsonnet++.o] エラー 1
----- end of file -----
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include=${opt-dir}/include
        --without-opt-include
        --with-opt-lib=${opt-dir}/lib
        --without-opt-lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/home/kenji/.local/share/mise/installs/ruby/3.3.1/bin/$(RUBY_BASE_NAME)
        --with-jsonnet-dir
        --without-jsonnet-dir
        --with-jsonnet-include=${jsonnet-dir}/include
        --without-jsonnet-include
        --with-jsonnet-lib=${jsonnet-dir}/lib
        --without-jsonnet-lib
        --use-system-libraries
/home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/mini_portile2-2.8.7/lib/mini_portile2/mini_portile.rb:623:in `block in execute': Failed to complete compile task (RuntimeError)
        from /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/mini_portile2-2.8.7/lib/mini_portile2/mini_portile.rb:589:in `chdir'
        from /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/mini_portile2-2.8.7/lib/mini_portile2/mini_portile.rb:589:in `execute'
        from extconf.rb:45:in `compile'
        from /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/mini_portile2-2.8.7/lib/mini_portile2/mini_portile.rb:237:in `cook'
        from extconf.rb:74:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3 for inspection.
Results logged to /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/extensions/x86_64-linux/3.3.0/jsonnet-0.5.3/gem_make.out
```

with `JSONNET_USE_SYSTEM_LIBRARIES=1`

```console
❯ env JSONNET_USE_SYSTEM_LIBRARIES=1 gem i jsonnet
Building native extensions. This could take a while...
ERROR:  Error installing jsonnet:
        ERROR: Failed to build gem native extension.

    current directory: /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3/ext/jsonnet
/home/kenji/.local/share/mise/installs/ruby/3.3.1/bin/ruby extconf.rb
checking for libjsonnet.h... yes
checking for -ljsonnet... yes
checking for libjsonnet_fmt.h... yes
creating Makefile

current directory: /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3/ext/jsonnet
make DESTDIR\= sitearchdir\=./.gem.20240813-783927-cp4job sitelibdir\=./.gem.20240813-783927-cp4job clean

current directory: /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3/ext/jsonnet
make DESTDIR\= sitearchdir\=./.gem.20240813-783927-cp4job sitelibdir\=./.gem.20240813-783927-cp4job
compiling callbacks.c
callbacks.c: In function ‘rescue_callback’:
callbacks.c:61:13: warning: function ‘rescue_callback’ might be a candidate for ‘gnu_printf’ format attribute [-Wsuggest-attribute=format]
   61 |             msg = rb_vsprintf(fmt, ap);
      |             ^~~
callbacks.c: In function ‘vm_set_import_callback’:
callbacks.c:152:37: error: passing argument 2 of ‘jsonnet_import_callback’ from incompatible pointer type [-Wincompatible-pointer-types]
  152 |     jsonnet_import_callback(vm->vm, import_callback_entrypoint, vm);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                     |
      |                                     char * (*)(void *, const char *, const char *, char **, int *)
In file included from callbacks.c:3:
/usr/include/libjsonnet.h:173:75: note: expected ‘int (*)(void *, const char *, const char *, char **, char **, size_t *)’ {aka ‘int (*)(void *, const char *, const char *, char **, char **, long unsigned int *)
’} but argument is of type ‘char * (*)(void *, const char *, const char *, char **, int *)’
  173 | void jsonnet_import_callback(struct JsonnetVm *vm, JsonnetImportCallback *cb, void *ctx);
      |                                                    ~~~~~~~~~~~~~~~~~~~~~~~^~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make: *** [Makefile:248: callbacks.o] エラー 1

make failed, exit code 2

Gem files will remain installed in /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/gems/jsonnet-0.5.3 for inspection.
Results logged to /home/kenji/.local/share/mise/installs/ruby/3.3.1/lib/ruby/gems/3.3.0/extensions/x86_64-linux/3.3.0/jsonnet-0.5.3/gem_make.out
```

--- a/ext/jsonnet/callbacks.c
+++ b/ext/jsonnet/callbacks.c
@@ -104,9 +104,15 @@ rubyjsonnet_jump_tag(const char *exc_mesg)
     return 0;
 }
 
+#ifdef HAVE_JSONNET_IMPORT_CALLBACK_0_19
+static int
+import_callback_entrypoint(void *ctx, const char *base, const char *rel, char **found_here,
+			   char **buf, size_t *buflen)
+#else
 static char *
 import_callback_entrypoint(void *ctx, const char *base, const char *rel, char **found_here,
 			   int *success)
+#endif
 {
     struct jsonnet_vm_wrap *const vm = (struct jsonnet_vm_wrap *)ctx;
     int state;
@@ -123,14 +129,26 @@ import_callback_entrypoint(void *ctx, const char *base, const char *rel, char **
 
     if (state) {
 	VALUE msg = rescue_callback(state, "cannot import %s from %s", rel, base);
+#ifdef HAVE_JSONNET_IMPORT_CALLBACK_0_19
+	*buf = rubyjsonnet_str_to_cstr(vm->vm, msg);
+	*buflen = RSTRING_LEN(msg);
+	return 1;
+#else
 	*success = 0;
 	return rubyjsonnet_str_to_cstr(vm->vm, msg);
+#endif
     }
 
     result = rb_Array(result);
-    *success = 1;
     *found_here = rubyjsonnet_str_to_cstr(vm->vm, rb_ary_entry(result, 1));
+#ifdef HAVE_JSONNET_IMPORT_CALLBACK_0_19
+    *buf = rubyjsonnet_str_to_cstr(vm->vm, rb_ary_entry(result, 0));
+    *buflen = RSTRING_LEN(rb_ary_entry(result, 0));
+    return 0;
+#else
+    *success = 1;
     return rubyjsonnet_str_to_cstr(vm->vm, rb_ary_entry(result, 0));
+#endif
 }
 
 /*

--- a/ext/jsonnet/extconf.rb
+++ b/ext/jsonnet/extconf.rb
@@ -13,8 +13,8 @@ unless using_system_libraries?
   require 'mini_portile2'
   message "Using mini_portile version #{MiniPortile::VERSION}\n"
 
-  recipe = MiniPortile.new('jsonnet', 'v0.18.0')
-  recipe.files = ['https://github.com/google/jsonnet/archive/v0.18.0.tar.gz']
+  recipe = MiniPortile.new('jsonnet', 'v0.20.0')
+  recipe.files = ['https://github.com/google/jsonnet/archive/v0.20.0.tar.gz']
   class << recipe
     CORE_OBJS = %w[
       desugarer.o formatter.o lexer.o libjsonnet.o parser.o pass.o static_analysis.o string_utils.o vm.o
@@ -89,4 +89,21 @@ end
 abort 'libjsonnet.h not found' unless have_header('libjsonnet.h')
 abort 'libjsonnet not found' unless have_library('jsonnet')
 have_header('libjsonnet_fmt.h')
+
+import_callback_0_19 = checking_for checking_message('JsonnetImportCallback >= v0.19.0') do
+  try_compile(<<SRC, '-Werror=incompatible-pointer-types')
+#include <libjsonnet.h>
+
+int f(void *ctx, const char *base, const char *rel, char **found_here, char **buf, size_t *buflen);
+
+int main() {
+  jsonnet_import_callback(NULL, f, NULL);
+  return 0;
+}
+SRC
+end
+if import_callback_0_19
+  $defs.push('-DHAVE_JSONNET_IMPORT_CALLBACK_0_19')
+end
+
 create_makefile('jsonnet/jsonnet_wrap')

